### PR TITLE
`USE_PRECOMPILED_HEADERS` is not supported on Apple M1

### LINKED ISSUE #80018: `USE_PRECOMPILED_HEADERS` is not supported on Apple M1
### 🐛 Describe the bug

I set `USE_PRECOMPILED_HEADERS` on and built PyTorch from source, errors occurred due to the following lines:

`error: Objective-C was disabled in PCH file but is currently enabled`

https://github.com/pytorch/pytorch/blob/7751ed41a6f81041c7eb0b21f45ac96a8bebe7dd/caffe2/CMakeLists.txt#L838-L840

### Versions

```
PyTorch version: 1.13.0a0+git44764f1
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] mypy==0.950
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torch==1.13.0a0+git44764f1
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torch                     1.13.0a0+git44764f1           dev_0    <develop>
```

cc @malfet @seemethere

### PR DESCRIPTION
**Pull request by TachikakaMin**
Fixes #80018

```bash
MACOSX_DEPLOYMENT_TARGET=12.6 CC=gcc CXX=g++ DEBUG=1 USE_DISTRIBUTED=0 USE_MKLDNN=0 USE_CUDA=0 BUILD_TEST=0 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 USE_XNNPACK=0 USE_PRECOMPILED_HEADERS=1 USE_MPS=1 python setup.py develop
```

`error: Objective-C was disabled in PCH file but is currently enabled`

This PR(https://github.com/pytorch/pytorch/pull/80432) has been reverted.

### CODE REVIEW

**Comment by TachikakaMin** [comment]
@albanD @peterbell10

**Comment by TachikakaMin** [comment]
@peterbell10 @kulinseth

Hi all, could you check this PR? Thanks.

**Comment by kulinseth** [comment]
@TachikakaMin why was the earlier change of `OBJCXX` enabling in the language was disabled?

**Comment by TachikakaMin** [comment]
@kulinseth 

Some tests failed. https://github.com/pytorch/pytorch/pull/80432#issuecomment-1170366368

**Comment by kulinseth** [comment]
The change looks fine to me.

**Review by kulinseth — APPROVED** [review]

**Comment by kulinseth** [comment]
@malfet, looks like the previous PR got reverted due to some failures. What testing should we enable on this PR to catch any breakages upfront ?
https://github.com/pytorch/pytorch/pull/80432#issuecomment-1170366368

**Comment by qqaatw** [comment]
Could you rebase and see if the failures are gone?

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -829,6 +829,15 @@ endif()
 set_property(SOURCE ${ATen_CORE_SRCS} APPEND
     PROPERTY COMPILE_DEFINITIONS "TORCH_ASSERT_ONLY_METHOD_OPERATORS")
 
+if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+  if(USE_MPS)
+    enable_language(OBJC OBJCXX)
+    foreach(input_filename ${Caffe2_MPS_SRCS})
+      set_source_files_properties(${input_filename} PROPERTIES COMPILE_FLAGS "-DUSE_MPS=1")
+    endforeach(input_filename)
```
Comment (qqaatw) [inline caffe2/CMakeLists.txt:837]: Do we need this?

`USE_MPS` is conditioned on `MPS_FOUND` by default
https://github.com/pytorch/pytorch/blob/e9be0b0fb93c271ba3deb31b5ccc8f4c4f602ef9/CMakeLists.txt#L229-L231

and `-DUSE_MPS=1` flag is added here:

https://github.com/pytorch/pytorch/blob/e9be0b0fb93c271ba3deb31b5ccc8f4c4f602ef9/CMakeLists.txt#L971-L972

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -789,6 +789,12 @@ setting `USE_PRECOMPILED_HEADERS=1` either on first setup, or in the
 USE_PRECOMPILED_HEADERS=1 python setup.py develop
 ```
 
+If you are using MacOS on Apple M1:
+
+```sh
+USE_PRECOMPILED_HEADERS=1 USE_MPS=1 python setup.py develop
```
Comment (malfet) [inline CONTRIBUTING.md:795]: One does not need to pass `USE_MPS` explicitly, it's enabled by default if OS supports it.
```suggestion
USE_PRECOMPILED_HEADERS=1 python setup.py develop
```

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -829,6 +829,15 @@ endif()
...
+  endif()
+endif()
```
Comment (malfet) [inline caffe2/CMakeLists.txt:839]:
```suggestion
if(USE_MPS OR USE_PYTORCH_METAL)
  enable_language(OBJC OBJCXX)
endif()
```

**Comment by qqaatw** [comment]
@TachikakaMin Hi, would you mind updating this PR per the review comments so that we can get it merged? Or if you don't mind, I can update it myself when I'm free. Thanks a lot!

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (TachikakaMin) [reply caffe2/CMakeLists.txt]: I don't remember why I do that. I just remembered if I don't add DUSE_MPS there is no such argument during compilation. I think the new commit is good.

**Comment by TachikakaMin** [comment]
@qqaatw Thanks! Just add the new commit.

**Comment by qqaatw** [comment]
@TachikakaMin Thanks@ Have you tested whether we can use PCH on macOS with the updated commits?

**Comment by TachikakaMin** [comment]
@malfet @qqaatw  I have tested the new patch. It isn't work. Some .mm files use aten/src/ATen/native/DispatchStub.h 

```python
// NB: This macro must be used in an actual 'cu' file; if you try using
// it from a 'cpp' file it will not work!
#if defined(__CUDACC__)
#define REGISTER_DISPATCH(name, fn) REGISTER_CUDA_DISPATCH(name, fn)
#elif defined(__HIPCC__)
// TODO: cut this over to HIP dispatch once we stop pretending that CUDA
// is HIP in the PyTorch HIPify build.
#define REGISTER_DISPATCH(name, fn) REGISTER_CUDA_DISPATCH(name, fn)
// #define REGISTER_DISPATCH(name, fn) REGISTER_HIP_DISPATCH(name, fn)
#elif defined(__OBJC__) && defined(USE_MPS)
// NB: this macro must be used from a 'mm' file in order to dispatch a MPS kernel
#define REGISTER_DISPATCH(name, fn) REGISTER_MPS_DISPATCH(name, fn)
#elif defined(CPU_CAPABILITY)
#define REGISTER_DISPATCH(name, fn) REGISTER_ARCH_DISPATCH(name, CPU_CAPABILITY, fn)
#define REGISTER_NO_AVX512_DISPATCH(name)       \
  REGISTER_AVX512_DISPATCH(name, nullptr)
#endif
```
defined(__OBJC__) is true but defined(USE_MPS) is false.

use the previous version can solve this problem:

```python
if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
  if(USE_MPS)
    enable_language(OBJC OBJCXX)
    foreach(input_filename ${Caffe2_MPS_SRCS})
      set_source_files_properties(${input_filename} PROPERTIES COMPILE_FLAGS "-DUSE_MPS=1")
    endforeach(input_filename)
  endif()
endif()
```

**Comment by peterbell10** [comment]
I think there might be a `CMAKE_OBJCXX_FLAGS` variable that needs to be synced with `CMAKE_CXX_FLAGS`
https://github.com/pytorch/pytorch/blob/4582ceb2c4a7316f09afd471a97d910347d21f01/CMakeLists.txt#L974

**Comment by TachikakaMin** [comment]
@peterbell10 @qqaatw @malfet

I think this PR can be merged.

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**